### PR TITLE
Parse parts, continuation and comments in string literals

### PR DESCRIPTION
**Pull request by mscfd**
I have worked on the idea suggested in #193 to solve the comment within string literal problem and also to identify parts of the string. I already experimented a bit with the extended tree using it for syntax highlighting and indentation of multiline string.

A string_literal is now parsed as a sequence like

```
(string_literal "\"" (string_literal_part) "\\" (string_literal) &
                (comment)  (comment) & (string_literal_part) "\"")
```

Double (double or single) quotes are parsed with escape node `"\\"` for the first quote character, and `(string_literal_part)` representing the second quote character. I am not convinced to really use `"\\"` as node name for the escape quote, as I sometimes struggle with whether to use one, two or four backslashes. But I would like to keep a separate anonymous or named node type for this, as it makes identifying escaped quotes easy. 

The true content of a string_literal can be constructed as concatenation of all `(string_literal_part)` children. Including `"\\"` in the concatenation as well yields the content as found in the fortran source.

The only quirk is that a string like `"012&abc"` needs to be broken into two `(string_literal_part)`, as the look-ahead and commit mechanics of the scanner dictates to emit a first `(string_literal_part)` for "012" and then use possibly extensive look-ahead to decide whether the ampersand is a continuation symbol.

The commit adds four new external tokens, a new scanner state `in_string` to track whether the scanner is within a string, with a continuation within a string and escaped quotes.

Examples can be found in the commit in the corpus. Unfortunately, the files in the corpus requires rather extensive updates, as any `(string_literal)` now becomes at least `(string_literal (string_literal_part))` and thus needs to be updated.

_Edit: Some fixes and quotes added._

### CODE REVIEW

**Comment by mscfd** [comment]
After thinking a bit about whether strings like "012&abc" can be scanned in one part, I found a solution. So that removes the quirk having to split content at ampersands, which are not continuation symbols, to organise the look-ahead.

The new commit was generated with 0.25, and thus fails the ckecks. Tests have been updated (one string_literal_part instead of two, were & was content) and run as before.

**Comment by mscfd** [comment]
The commit message was edited.

**Comment by mscfd** [comment]
Parser with treesitter 0.26 generated.

**Comment by mscfd** [comment]
I have gone ahead and implemented support for this extension in f90-ts-mode (emacs). So far, it works great [PR string_literal](https://github.com/mscfd/emacs-f90-ts-mode/pull/170). It improves syntax highlighting, indentation with several strategies and break/join and fill operations. The emacs mode has a number of ert based tests, which also indirectly test the grammar.

File: grammar.js [hunk 1]
```diff
@@ -2219,7 +2224,14 @@ module.exports = grammar({
         // also need to *capture* it here
         token.immediate('_'),
       )),
-      $._string_literal,
+      alias($.string_literal_start, '"'),
```
Comment (ZedThree) [inline grammar.js:2227]: I am a little bit worried that this might be a bit confusing, because this will also capture single quotes. Of course, the actual quote character can be recovered from the node text, so it's not the end of the world by any means.

What if `string_literal_start` was just an unnamed node? (that is, just call it `_string_literal_start`) How does that change things?

File: test/corpus/expressions.txt [hunk 2]
```diff
@@ -161,46 +163,76 @@ END PROGRAM
     (program_statement
       (name))
     (assignment_statement
-      left: (identifier)
-      right: (string_literal))
+      (identifier)
```
Comment (ZedThree) [inline test/corpus/expressions.txt:166]: Not a big deal, but could you add back the field names?

**Review by ZedThree — CHANGES_REQUESTED** [review]
To be honest, I was quite worried about this making string literals always having at least one child node, but looking at some other language grammars (C, for example), they also do something similar, having `(string_literal (string_content))`, which they need to capture implicit literal concatenation.

This definitely makes it much better in terms of being able to capture the non-string bits for continued literals.

Maybe it also makes sense to add a CST test, to check that it's capturing the anonymous nodes correctly too?

File: grammar.js (hunk 1 above)
Comment (mscfd) [reply grammar.js:2227]: With an unnamed node like _string_literal_start, those nodes disappear and are not available for queries in syntax highlight etc.
My reasoning was (as far as I remember), that node type and text are always strictly separated. Similar to using the case insensitive keyword string as an alias. Type and text look similar, except the case has changed.
I thought about using `quote` instead of `"`, but felt like a whole word does not fit the character of the grammar, considering that other symbols are also stored with types like `=`. But if this feels confusing, a word like quote would give for example:
```
2:7  - 2:12         right: string_literal 
2:7  - 2:8            "quote"
2:8  - 2:11           string_literal_part `abc`
2:11 - 2:12           "quote"
```

File: test/corpus/expressions.txt (hunk 2 above)
Comment (mscfd) [reply test/corpus/expressions.txt:166]: Not sure how and when that got lost. But I struggled with the cst. I wanted to add a cst test, but was not able. I will try once more.

File: grammar.js (hunk 1 above)
Comment (ZedThree) [reply grammar.js:2227]: Ok, that seems reasonable

**Comment by DisciplinedSoftware** [comment]
@mscfd I've prepared a follow-up addressing @ZedThree's two requests (restoring the accidentally removed corpus field names, and adding a `:cst` test for the anonymous string delimiters, escaped-quote marker, and continuation markers). It's a single commit based directly on this PR's current head:

- PR against your `string_literal` branch: https://github.com/mscfd/tree-sitter-fortran/pull/1
- Commit: DisciplinedSoftware/tree-sitter-fortran@b250ea3a779e10076e4e8e4235809277059597a3

Alternatively, cherry-pick it directly:
```sh
git fetch https://github.com/DisciplinedSoftware/tree-sitter-fortran.git pr198-string-literal-review
git cherry-pick b250ea3a779e10076e4e8e4235809277059597a3
```

Passes `tree-sitter test` and `npx tree-sitter-cli@0.26.3 test`. Merging or cherry-picking it will update this PR automatically.

File: src/scanner.c [hunk 3]
```diff
@@ -379,69 +402,205 @@ static bool scan_string_literal_kind(TSLexer *lexer) {
...
+// It consumes any trailing blanks and decides whether the '&' is a line
+// continuation marker or ordinary string content.
+// It does not set and advance the end marker.
+static bool string_ampersand_is_continuation(Scanner *scanner, TSLexer *lexer) {
+    while (iswblank(lexer->lookahead)) {
+        advance(lexer);
+    }
+    return (lexer->lookahead == '\n' || lexer->lookahead == '\r' ||
+            lexer->lookahead == '!'  || lexer->eof(lexer));
+}
```
Comment (DisciplinedSoftware) [inline src/scanner.c:439]: **Regression: `!` after `&` is classified as a continuation inside character context.**

This predicate returns true when `&`, optional blanks, and then `!` are seen. Inside a string literal that is valid content, not commentary — the standard does not allow a comment following a continuation `&` in character context, so `& !b"` here must remain literal text.

Reproducer (parses cleanly on `master`, cascading `ERROR` on this branch because `!b"` is consumed as a comment and the string never closes):

```fortran
print *, "a & !b"
print *, "x"
end
```

Suggestion: drop the `lexer->lookahead == '!'` case from this check (continuation in a string should only be `&` + blanks + newline/EOF), or if you want to keep it, first verify that no closing quote follows later on the same line before treating the `!` as a comment.

File: src/scanner.c [hunk 4]
```diff
@@ -379,69 +402,205 @@ static bool scan_string_literal_kind(TSLexer *lexer) {
...
+
+    bool has_content = false;
+
+    while (!lexer->eof(lexer)) {
+        if (lexer->lookahead == '\n' || lexer->lookahead == '\r') {
+            // unterminated string line: only valid if we have accumulated
+            // content, the grammar will hopefully close the string via error
+            // recovery
+            lexer->result_symbol = STRING_LITERAL_PART;
+            return has_content;
```
Comment (DisciplinedSoftware) [inline src/scanner.c]: **Regression: `in_string` is left set after a bare newline terminates an unterminated literal.**

On a newline without a preceding `&`, this emits the accumulated `STRING_LITERAL_PART` but does not reset `scanner->in_string` / `string_quote`. On the next call, `scan()` skips the newline (lines 751–768) and then dispatches straight back into `scan_string_literal_content`, so the following source lines are consumed as string content until another matching quote is found.

Reproducer:

```fortran
x = "abc
y = 1
z = 2
print *, "ok"
end
```

On `master` the damage is line-local (`x = "abc` errors, the remaining statements parse). On this branch most of the file collapses into a single `ERROR`/string. This matters a lot for editor use, where a momentarily missing quote is the common case.

Suggestion: when a bare `\n`/`\r` is hit here, reset `scanner->in_string = IN_STRING_NONE` (and `string_quote`) before returning, so recovery stays line-local. The explicit `&` continuation path in `scan_string_ampersand` should be unaffected since it returns before reaching this branch.

File: src/scanner.c [hunk 5]
```diff
@@ -612,9 +771,18 @@ static bool scan(Scanner *scanner, TSLexer *lexer, const bool *valid_symbols) {
...
     }
 
-    if (valid_symbols[STRING_LITERAL]) {
-        if (scan_string_literal(lexer)) {
-            return true;
+    // skip string parsing if we are in a line continuation state
+    if (!scanner->in_line_continuation) {
+        if (scanner->in_string == IN_STRING_NORMAL || scanner->in_string == IN_STRING_QUOTE_QUOTE) {
+            if (valid_symbols[STRING_LITERAL_PART] || valid_symbols[STRING_LITERAL_QUOTE]) {
+                if (scan_string_literal_content(scanner, lexer)) {
```
Comment (DisciplinedSoftware) [inline src/scanner.c]: **Leading blanks are dropped from `string_literal_part`.**

`scan()` unconditionally skips blanks at lines 751–754 (and again at 766–768) before reaching this in-string dispatch. Because a new external-scanner call happens at every token boundary inside the string (after the start quote, after a `&` continuation, after a doubled quote), any blanks at those boundaries are silently discarded and never appear in a part node. That contradicts the stated goal that concatenating the part children reproduces the literal's content.

Concrete cases:

- `x = "  abc"` → part is `abc` (two blanks lost)
- `x = "abc &\n    & def"` → the blank after the second `&` is lost
- `x = "a"" b"` → the blank after the doubled quote is lost
- `x = "   "` → no part node at all

Suggestion: when `scanner->in_string != IN_STRING_NONE` (and not in line continuation), dispatch to `scan_string_literal_content` *before* the generic blank skipping, or otherwise arrange for those blanks to be included in the following part.

**Review by DisciplinedSoftware — CHANGES_REQUESTED** [review]
Thanks for this work. Splitting string literals into start/part/quote/end nodes is a promising structural approach, and the regenerated artifacts and corpus tests look consistent with the grammar changes.

However, while exercising the scanner against the base I found three behavioral regressions in `src/scanner.c` that I think need correcting before this can be approved. Each is detailed inline with a minimal reproducer:

1. **`!` treated as continuation inside a string** — valid content like `"a & !b"` now fails to parse.
2. **`in_string` not reset on a bare newline** — an unterminated literal now swallows subsequent lines until the next matching quote, badly degrading error recovery.
3. **Leading blanks dropped from `string_literal_part`** — the generic blank skipping in `scan()` runs before the in-string dispatch, so concatenating parts no longer reconstructs the true content.

The field-name / CST-test feedback already raised is being handled separately in mscfd/tree-sitter-fortran#1, so I have intentionally left it out here.

**Comment by mscfd** [comment]
@DisciplinedSoftware: Thanks a lot for spotting the problems and fixing the corpus. I will cherry-pick the commit. If you like, and as you seem to have a good idea how to fix them, I can also cherry-pick fixes for the three wrongly parsed corner-cases you found. I probably need a few days to find time for that.

**Comment by DisciplinedSoftware** [comment]
@mscfd Thanks — here are the fixes for the three corner cases, one commit per fix with a corpus test each, based on `b250ea3` so they cherry-pick cleanly in order:

- PR against your `string_literal` branch: https://github.com/mscfd/tree-sitter-fortran/pull/2
- Commits: `0f2044a` (`& !` inside a string is content), `8f88795` (leading blanks kept in parts), `2ba313f` (unterminated literal ends with the line)

Or cherry-pick directly:
```sh
git fetch https://github.com/DisciplinedSoftware/tree-sitter-fortran.git pr198-scanner-fixes
git cherry-pick 0f2044af4acebcf5ba1a7f83de939a57a3696f16 8f887956be8848a4155a5d34363960bca459c310 2ba313f62c88079bf74585a774e88a7efd3b0052
```

Two things worth a look:

- The first commit drops `str5` from the "Line continuation in concat string" test (`"...ijk&  ! comment` continued on the next line). gfortran and ifx both reject that form with "Unterminated character constant", since in a character context the `&` may not be followed by commentary, and keeping it as a continuation is what broke `"a & !b"`. The split doubled quote with a comment (`"abc"&  ! comment` continued by `&"def"`) is kept as it is today — `scan_string_quote_ampersand` has already looked past the comment there, and gfortran reads it as `abc"def` too — and is now covered by a test.
- The third commit emits the `string_literal_part` at a bare newline even when it is empty (a zero-width token). That is needed because tree-sitter only stores scanner state together with a token; without it the state of the opening quote is restored on the next call and the next line is scanned as content. It only shows up in the ERROR of an unterminated literal.

Scanner only, no grammar change, so `parser.c` is untouched. Passes `tree-sitter test`, `npx tree-sitter-cli@0.26.3 test`, `cargo test` and `examples/*` at every commit.

Two smaller things I noticed on the way, outside the three regressions — happy to send commits for either if you want them:

- `$._string_literal` is still declared in `externals` in `grammar.js`, and `STRING_LITERAL` in the scanner's `TokenType`, but the scanner never emits it any more; both could go (this one needs a `parser.c` regeneration, which is why I left it out of the cherry-picks).
- `scan_string_quote_ampersand` accepts commentary after the `&` when it looks ahead for a split doubled quote. That matches gfortran (it compiles `"abc"&  ! comment` / `&"def"` to `abc"def` without even a warning), while ifx follows the standard and reads `"abc"` as closed, so the tree differs from what ifx would accept for that input. I left the behaviour as you designed it; a comment in the code saying it follows gfortran would make the choice explicit.

_Edit: the commit ids above were updated after folding the split-doubled-quote exception into the first commit._

**Comment by mscfd** [comment]
Thanks @DisciplinedSoftware. All four commits are now on branch mscfd:string_literal, but I have not yet looked into them.

**Comment by ZedThree** [comment]
I've just been looking at the CST of some of the new examples, and I found this one quite confusing:

```f90
  s = 'abc'&  ! a continuation of a doubled quote
      &'def'
```

```
6:2  - 7:12       assignment_statement
6:2  - 6:3          left: identifier `s`
6:4  - 6:5          "="
6:6  - 7:12         right: string_literal
6:6  - 6:7            "\""
6:7  - 6:10           string_literal_part `abc`
6:10 - 6:11           "\\"
6:11 - 6:12           "&"
6:14 - 6:49           comment `! a continuation of a doubled quote`
7:6  - 7:7            "&"
7:7  - 7:8            string_literal_part `'`
7:8  - 7:11           string_literal_part `def`
7:11 - 7:12           "\""
```

but I think I understand it now: escaped quotes are now always represented by `("\\") (string_literal_part <quote_kind>)`, and this is consistent regardless of whether or not the escaped quote is split up or not.

The downside is that using them normally is now a bit noisier:

```f90
s = 'abc''def'
```
```
9:2   - 9:16        assignment_statement
9:2   - 9:3           left: identifier `s`
9:4   - 9:5           "="
9:6   - 9:16          right: string_literal
9:6   - 9:7             "\""
9:7   - 9:10            string_literal_part `abc`
9:10  - 9:11            "\\"
9:11  - 9:12            string_literal_part `'`
9:12  - 9:15            string_literal_part `def`
9:15  - 9:16            "\""
````

whereas before it would've just been 

```
9:6   - 9:16          right: string_literal `'abc''def'`
```

@mscfd This is your intended behaviour, right?

**Comment by mscfd** [comment]
@ZedThree: Your understanding is correct. I chose to make the first quote an unnamed node of type "\" (or "\\" in most non-fortran) and the second quote a separate string_literal_part. First, it allows easy syntax-highlighting of the first quote (like with escape-face in emacs) and easy access to the second quote (by querying the next child of type string_literal_part. Furthermore, depending on whether one concatenates nodes of type `string_literal_part` or `string_literal_part` and `\\` gets the true or the fortran string content. Other operations like break or join of lines also become easier if the structure, as allowed by the standard, is exposed. So independently of continuation, I always want to expose the structural details to ease usage of the CST. For editor usage, that additional information is really helpful. And if you just want to recover the whole string, then the node `string_literal` serves that purpose just as before.

Btw: Your example seems to be invalid (even if accepted by gfortran), as @DisciplinedSoftware mentioned. Not sure what to do about that, but I would not accept it. Or is this used in real code anywhere? Getting an ERROR node can be helpful (for example, I added error node overlays in emacs, and they are a simple and helpful feedback).

**Comment by mscfd** [comment]
@DisciplinedSoftware: I have checked your commits and they all look correct and and are surprisingly simple and straightforward fixes. The CST in the corpus also look as expected.
And the deserialization was indeed missing for the string literal state variables.

**Comment by ZedThree** [comment]
> Btw: Your example seems to be invalid (even if accepted by gfortran), as @DisciplinedSoftware mentioned. Not sure what to do about that, but I would not accept it. Or is this used in real code anywhere? Getting an ERROR node can be helpful (for example, I added error node overlays in emacs, and they are a simple and helpful feedback).

[Unfortunately, several compilers accept it](https://godbolt.org/z/5qjKWf3Ke), although I agree that the standard is pretty clear:

> If a character context is to be continued, an “&” shall be the last nonblank character on the line.

I searched my corpus (~50M lines of code) and didn't find a single instance, so it must be pretty rare.

On the other hand, it already works so :shrug:

**Comment by mscfd** [comment]
It is such a niche feature/error that I personally do not care. And the scanner does not change much either way.

PS: I just updated string_literal branch to remove the unused external token string_literal, before this observation by @DisciplinedSoftware gets lost. The commit still needs to be updated with tree-sitter 0.26.
I also noticed that the cst test generation just does not work with the 0.25 variants, those tests just fail with weird diffs.

File: test/corpus/expressions.txt (hunk 2 above)
Comment (mscfd) [reply test/corpus/expressions.txt]: Resolved by cherry-picking commit from @DisciplinedSoftware

**Comment by mscfd** [comment]
Sorry, I got a bit confused: a comment after ampersand is (wrongly) accepted by gfortran and the current scanner, if it sits in-between an escaped-quote-sequence. We should keep it that way. No need to worry about this I think. @ZedThree: I missed that your example did exactly that and thought it had the comment after ampersand in a general string_literal_part context.

**Comment by mscfd** [comment]
Branch updated with tree-sitter 0.26.

**Review by DisciplinedSoftware — APPROVED** [review]
Re-reviewed at `907157e`; the three regressions I raised are fixed. I can't resolve review threads on this repo, so @mscfd, feel free to close my three `src/scanner.c` threads.

- The four cherry-picked commits (`b250ea3`, `0f2044a`, `8f88795`, `2ba313f`) are tree-identical to what I submitted in mscfd/tree-sitter-fortran#1 and #2.
- `907157e` (drop the unused `_string_literal` external): the `externals` list in `grammar.js` and `enum TokenType` in `src/scanner.c` stay in the same order (15 entries), nothing references the removed token, and `tree-sitter-cli@0.26.3 generate` reproduces the committed `src/parser.c` byte for byte.
- `tree-sitter-cli@0.26.3 test`: 147 passed, 0 failed. My three reproducers parse correctly at head: `"a & !b"` is a single 6-character part; an unterminated `"abc` errors line-locally and a later `"ok"` is a normal literal; `"  abc"` keeps its leading blanks.

Thanks @mscfd for picking these up.
